docs/_docs/04_reference/config-blocks-and-attributes.md

```diff
@@ -393,6 +394,63 @@ inputs = {
 }
 ```
 
+**What is deep merge?**
+
+When the `merge_strategy` for the `include` block is set to `deep`, Terragrunt will perform a deep merge of the included
+config. For Terragrunt config, deep merge is defined as follows:
+
+- For simple types, the source overrides the target.
+- For lists, the two attribute lists are combined together in concatenation.
+- For maps, the two maps are combined together recursively. That is, if the map keys overlap, then a deep merge is
+  performed on the map value.
+- For blocks, if the label is the same, the two blocks are combined together recursively. Otherwise, the blocks are
+  appended like a list. This is similar to maps, with block labels treated as keys.
+
+However, due to internal implementation details, some blocks are not deep mergeable. This will change in the future, but
+for now, terragrunt performs a shallow merge (that is, block definitions in the child completely override the parent
+definition). The following blocks have this limitation:
+    - `remote_state`
+    - `generate`
+
+Similarly, the `locals` block is deliberately omitted from the merge operation by design. That is, you will not be able
+to access parent config `locals` in the child config, and vice versa in a merge. However, you can access the parent
+locals in child config if you use the `expose` feature.
+
+Finally, `dependency` blocks have special treatment. When doing a `deep` merge, `dependency` blocks from **both** child
+and parent config are accessible in **both** places. For example, consider the following setup:
```

Comment:
Interesting. Why?

Reply:
I don't really have a good answer to this other than that it naturally happened when I coded the deep merge. I actually realized this fact when I started testing it.

I remember chasing the reasoning when I realized it during implementation, but it's been about 2 months since I coded this up so I have forgotten it now...